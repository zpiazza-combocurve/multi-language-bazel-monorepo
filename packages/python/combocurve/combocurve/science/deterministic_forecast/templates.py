from combocurve.science.forecast.auto_forecast_warnings import NO_WARNING

EMPTY_P_SEGS = {'best': {'segments': [], 'diagnostics': {}, 'eur': None, 'rur': None}}

EMPTY_RATIO = {'segments': [], 'diagnostics': {}, 'basePhase': None, 'x': None, 'eur': None, 'rur': None}


# Moved into its own file to avoid circular dependency error
def return_template(P_dict: dict = EMPTY_P_SEGS,
                    diagnostics: dict = {},
                    forecastType: str = 'rate',
                    forecastSubType: str = 'automatic',
                    warning: dict = NO_WARNING,
                    ratio: dict = EMPTY_RATIO,
                    forecasted: bool = True,
                    data_freq: str = None) -> dict:

    ret = {
        'P_dict': P_dict,
        'diagnostics': diagnostics,
        'forecastType': forecastType,
        'forecastSubType': forecastSubType,
        'warning': warning,
        'ratio': ratio,
        'forecasted': forecasted,
        'data_freq': data_freq
    }
    return ret
