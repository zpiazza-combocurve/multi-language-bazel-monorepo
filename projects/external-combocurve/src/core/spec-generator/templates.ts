export type ObjectDefinitionSpec = {
	name: string;
	properties: PropertyDefinitionSpec[];
	required?: string[];
	example?: string[];
};

export type PropertyDefinitionSpec = {
	name: string;
	type: string;
	typeKey: string;
	required?: boolean;
	isArray?: boolean;
	itemsTypeKey?: string;
	itemsType?: string;
	example?: string;
	stringEnum?: string[];
	minimum?: number;
	maximum?: number;
	minItems?: number;
	maxItems?: number;
	maxLength?: number;
	format?: string;
};

export const definitionsTemplate = `
{{#each objects}}
{{this.name}}:
  type: 'object'
  {{#if required}}
  required:
    {{#each required}}
    - {{this}}
	{{/each}}
  {{/if}}
  properties:
  {{#each properties}}
    {{this.name}}:
    {{#if isArray}}
      type: 'array'
      items:
        {{itemsTypeKey}}: '{{itemsType}}'
		{{#if example}}
        example: '{{example}}'
        {{/if}}
		{{#if maxItems}}
        maxItems: {{maxItems}}
        {{/if}}
		{{#if minItems}}
        minItems: {{minItems}}
        {{/if}}
	{{else}}
      {{typeKey}}: '{{type}}'
	  {{#if format}}
      format: '{{format}}'
      {{/if}}
	  {{#if example}}
      example: '{{example}}'
      {{/if}}
      {{#if stringEnum}}
      enum:
        {{#each stringEnum}}
        - '{{this}}'
	    {{/each}}
	  {{/if}}
	  {{#if minimum}}
      minimum: '{{minimum}}'
      {{/if}}
	  {{#if maximum}}
      maximum: '{{maximum}}'
      {{/if}}	    
    {{/if}}
  {{/each}}
  {{#if example}}
  example:
  {{#each example}}
    {{this}}	
  {{/each}}
  {{/if}}	
{{/each}}
`;

export type EndpointPathSpec = {
	verb: string;
	id?: string;
	description?: string;
	consumes?: string;
	produces?: string;
	parameters: EndpointParameterSpec[];
	responses: EndpointResponseSpec[];
};

export type EndpointParameterSpec = {
	fromWhere: string;
	name: string;
	type?: string;
	stringEnum?: string[];
	description?: string;
	format?: string;
	default?: string;
	minimum?: number;
	maximum?: number;
	required?: boolean;
	schema?: string;
	isArray?: boolean;
	itemsTypeKey?: string;
	itemsType?: string;
	example?: string;
	itemsExample?: string;
	minItems?: number;
	maxItems?: number;
};

export type EndpointResponseSpec = {
	status: number;
	description?: string;
	schema?: string;
	headers?: EndpointResponseHeaderSpec[];
};

export type EndpointResponseHeaderSpec = {
	name: string;
	type: string;
};

export const pathsTemplate = `
{{#each routes}}
{{completeRoute}}:
  {{#each methods}}
  {{verb}}:
    operationId: {{id}}
	{{#if description}}
    description: {{description}}
    {{/if}}
    {{#if consumes}}
    consumes:
      - {{consumes}}
    {{/if}}
    {{#if produces}}
    produces:
      - {{produces}}
    {{/if}}
    parameters:
    {{#each parameters}}
      - name: {{name}}
        in: {{fromWhere}}
		{{#if description}}
        description: {{description}}
        {{/if}}
        {{#if required}}
        required: true
        {{/if}}
		{{#if type}}
        type: {{type}}
        {{/if}}
		{{#if format}}
        format: {{format}}
        {{/if}}
		{{#if minimum}}
        minimum: {{minimum}}
        {{/if}}
		{{#if maximum}}
        maximum: {{maximum}}
        {{/if}}
		{{#if stringEnum}}
        enum:
        {{#each stringEnum}}
          - '{{this}}'
	    {{/each}}
	    {{/if}}
		{{#if example}}
        example: {{example}}
        {{/if}}        
		{{#if schema}}
        schema:
          $ref: '{{schema}}'
      	{{/if}}
	  {{#if isArray}}
		{{#if minItems}}
        minItems: {{minItems}}
      	{{/if}}
		{{#if maxItems}}
        maxItems: {{maxItems}}
		{{/if}}
        items:
          {{itemsTypeKey}}: {{itemsType}}
		  {{#if itemsExample}}
          example: {{itemsExample}}
          {{/if}}
	  {{/if}}
    {{/each}}
    responses:
    {{#each responses}}
      '{{status}}':
        description: {{description}}
        {{#if headers}}
        headers:
        {{#each headers}}
          {{name}}:
            type: {{type}}
            {{#if description}}
            description: >-
              {{description}}
            {{/if}}    
        {{/each}}
      {{/if}}
      {{#if schema}}
        schema:
          $ref: {{schema}}
      {{/if}}
    {{/each}}
    {{#if tags}}
    tags:
      {{#each tags}}
      - {{this}}
      {{/each}}
    {{/if}}
  {{/each}}
{{/each}}
`;
