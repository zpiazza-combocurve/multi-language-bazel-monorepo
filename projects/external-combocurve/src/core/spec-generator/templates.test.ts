import Handlebars from 'handlebars';

import { definitionsTemplate, pathsTemplate } from './templates';

describe('templates', () => {
	it('should generate definitions', () => {
		const pricingEconModelSample = {
			objects: [
				{
					name: 'PricingModelOil',
					properties: [
						{
							name: 'cap',
							type: 'number',
							typeKey: 'type',
						},
						{
							name: 'escalationModel',
							type: 'string',
							typeKey: 'type',
						},
						{
							name: 'rows',
							type: 'array',
							typeKey: 'type',
							isArray: true,
							itemsTypeKey: '$ref',
							itemsType: '#/definitions/PricingModelOilRow',
						},
					],
				},
				{
					name: 'PricingModelOilRow',
					properties: [
						{
							name: 'price',
							type: 'number',
							typeKey: 'type',
						},
						{
							name: 'dates',
							type: '#/definitions/DateCriteria',
							typeKey: '$ref',
						},
						{
							name: 'offsetToAsOf',
							type: '#/definitions/OffsetCriteria',
							typeKey: '$ref',
						},
						{
							name: 'entireWellLife',
							type: 'string',
							typeKey: 'type',
						},
					],
				},
			],
		};

		const template = Handlebars.compile(definitionsTemplate);
		const result = template(pricingEconModelSample);

		expect(result).toMatchSnapshot();
	});

	it('should generate paths', () => {
		const paths = {
			routes: [
				{
					completeRoute: '/v1/projects/{projectId}/econ-models/pricing',
					methods: [
						{
							verb: 'post',
							id: 'post-v1-pricing',
							consumes: 'application/json',
							produces: 'application/json',
							parameters: [
								{
									fromWhere: 'path',
									name: 'projectId',
									type: 'string',
									required: true,
								},
								{
									fromWhere: 'body',
									name: 'pricing',
									schema: '#/definitions/PricingInputList',
								},
							],
							responses: [
								{
									status: 207,
									description: 'Multi-Status',
									schema: '#/definitions/PricingMultiStatusResponse',
								},
							],
						},
					],
				},
			],
		};

		const template = Handlebars.compile(pathsTemplate);
		const result = template(paths);

		expect(result).toMatchSnapshot();
	});
});
