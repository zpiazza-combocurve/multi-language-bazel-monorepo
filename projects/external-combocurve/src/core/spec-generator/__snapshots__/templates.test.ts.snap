// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`templates should generate definitions 1`] = `
"
PricingModelOil:
  type: 'object'
  properties:
    cap:
      type: 'number'
    escalationModel:
      type: 'string'
    rows:
      type: 'array'
      items:
        $ref: '#/definitions/PricingModelOilRow'
PricingModelOilRow:
  type: 'object'
  properties:
    price:
      type: 'number'
    dates:
      $ref: '#/definitions/DateCriteria'
    offsetToAsOf:
      $ref: '#/definitions/OffsetCriteria'
    entireWellLife:
      type: 'string'
"
`;

exports[`templates should generate paths 1`] = `
"
/v1/projects/{projectId}/econ-models/pricing:
  post:
    operationId: post-v1-pricing
    consumes:
      - application/json
    produces:
      - application/json
    parameters:
      - name: projectId
        in: path
        required: true
        type: string
      - name: pricing
        in: body
        schema:
          $ref: '#/definitions/PricingInputList'
    responses:
      '207':
        description: Multi-Status
        schema:
          $ref: #/definitions/PricingMultiStatusResponse
"
`;
